torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -543,507 +473,32 @@ def aot_dispatch_autograd(
-                    CompiledFunction._compiled_autograd_key
-                )
-
-                # Pass args even though they're unused, so that the graph is built
-                out = CompiledFunctionBackward.apply(*all_args)
-            else:
-                out = call_compiled_backward()
-
-            # TODO: figure out how to refactor the backward properly so I can use aot_dispatch_subclass_wrapper() here.
-            if CompiledFunction.maybe_subclass_metadata is not None:
-                assert (
-                    CompiledFunction.maybe_subclass_metadata.grad_input_metas
-                    is not None
-                )
-                outs_wrapped = wrap_tensor_subclasses(
-                    out,
-                    subclass_metas=CompiledFunction.maybe_subclass_metadata.grad_input_metas,
-                )
-                return (*[None] * num_tokens, *outs_wrapped)
-            return (*[None] * num_tokens, *out)
+    lazy_backward_info = AutogradLazyBackwardCompileInfo(
+        bw_module,
+        placeholder_list,
+        saved_context,
+        saved_compile_context,
+    )
 
-    compiled_function = RuntimeWrapper(
-        indices_of_inps_to_detach=_indices_of_inps_to_detach,
-        trace_joint=True,
-        disable_amp=disable_amp,
-    ).post_compile(
-        CompiledFunction.apply,
+    compiled_function = AOTDispatchAutograd.post_compile(
+        compiled_fw_func,
+        compiled_bw_func,
+        maybe_subclass_meta,
+        num_symints_saved_for_bw,
+        backward_state_indices,
+        disable_amp,
+        _indices_of_inps_to_detach,
+        lazy_backward_info,
         aot_config,
-        runtime_metadata=fw_metadata,
+        fw_metadata=fw_metadata,
     )
 
-    if not config.debug_assert:
-        return compiled_function
-
-    flat_requires_grad = [
-        a.requires_grad if isinstance(a, Tensor) else None for a in flat_args
-    ]
-
-    @wraps(compiled_function)
-    def debug_compiled_function(args: List[Any]):
-        # TODO: Check aliasing relationships
-        # TODO: Check strides for metadata mutation
-        # (NB: ideally, this logic is factored out of this function and
-        # you move these debug checks there)
-
-        # Check requires grad.  Bad case is when we compiled with
-        # requires_grad = False, but input requires_grad = True
-        # (vice versa is OK; we compute a gradient and then throw
-        # it away when it hits the input.)
-        for i, a in enumerate(args):
-            can_require_grad = flat_requires_grad[i]
-            if can_require_grad is None:
-                assert not isinstance(a, Tensor)
-            elif not can_require_grad:
-                assert not a.requires_grad, format_guard_bug_msg(
-                    aot_config,
-                    f"{describe_input(i, aot_config)} would not require grad",
-                )
-
-        return compiled_function(args)
+    if config.debug_assert:
```

Comment:
It does feel a bit uneccessary to include config-driven debug assert logic in the cache output, especially if it adds a lot of complexity - if you're running debug asserts, you're most likely hitting a (rare) compile bug (since we have this config off by default).

I think I would base the decision more on complexity: if you think including including the debug assert runtime wrapper in the caching logic isn't a major complexity burden, then all-else-equal it probably is nice to include (that way if someone happens to turn on this config, they don't need to worry about turning off caching to get it to work)